File: internal/services/container/nodekind/export.go
package nodekind

import (
	"encoding/json"
	"sort"
)

// ExportedSpec is the JSON-friendly view of Spec — strips function pointers
// (ExecOutFn, DataInDynamicFn) since they can't be compared across languages,
// keeps only static metadata. Both backend and frontend dump this shape to
// JSON; the D2 parity test diffs the two outputs to enforce no drift.
//
// Excludes IsYield (validator-only on backend, no frontend equivalent) — D2
// parity test skips this field.
//
// Includes HasExecOutFn / HasDataInDynamicFn booleans so parity can assert
// "both sides agree this kind has a dynamic-pin function" without comparing
// the function bodies (which would require cross-language semantic analysis).
type ExportedSpec struct {
	Kind               string             `json:"kind"`
	Group              string             `json:"group"`
	ExecIn             []string           `json:"execIn"`
	ExecOut            []string           `json:"execOut"` // static only; dynamic via execOutFn
	HasExecOutFn       bool               `json:"hasExecOutFn"`
	DataIn             map[string]PinType `json:"dataIn"`
	HasDataInDynamicFn bool               `json:"hasDataInDynamicFn"`
	DataOut            map[string]PinType `json:"dataOut"`
	IsPureData         bool               `json:"isPureData"`
	IsVisualOnly       bool               `json:"isVisualOnly"`
}

// ExportJSON returns all registered specs as a sorted JSON array (stable
// output for diffing with frontend export). The slice is sorted by Kind
// ascending so two dumps from semantically-equal registries produce
// byte-identical output.
func ExportJSON() ([]byte, error) {
	specs := make([]ExportedSpec, 0, len(byKind))
	for _, s := range byKind {
		specs = append(specs, ExportedSpec{
			Kind:               s.Kind,
			Group:              s.Group,
			ExecIn:             normalizeStringSlice(s.ExecIn),
			ExecOut:            normalizeStringSlice(s.ExecOut),
			HasExecOutFn:       s.ExecOutFn != nil,
			DataIn:             normalizePinMap(s.DataIn),
			HasDataInDynamicFn: s.DataInDynamicFn != nil,
			DataOut:            normalizePinMap(s.DataOut),
			IsPureData:         s.IsPureData,
			IsVisualOnly:       s.IsVisualOnly,
		})
	}
	sort.Slice(specs, func(i, j int) bool { return specs[i].Kind < specs[j].Kind })
	return json.MarshalIndent(specs, "", "  ")
}

// normalizeStringSlice returns nil for empty/nil slices so JSON output is
// consistently `null` (not `[]`) — matches frontend's choice of `null` for
// empty arrays, ensuring byte-identical diffs.
func normalizeStringSlice(s []string) []string {
	if len(s) == 0 {
		return nil
	}
	return s
}

// normalizePinMap returns nil for empty/nil maps for the same reason as
// normalizeStringSlice: byte-identical JSON output for empty maps.
func normalizePinMap(m map[string]PinType) map[string]PinType {
	if len(m) == 0 {
		return nil
	}
	return m
}

File: frontend/src/components/containers/nodeRegistry/export.ts
// frontend/src/components/containers/nodeRegistry/export.ts
//
// Mirror of backend internal/services/container/nodekind/export.go ExportJSON.
// Both sides produce byte-identical JSON for the same set of registered specs
// (modulo the fields each side carries — IsYield is BE-only).
//
// D2 parity test consumes this: a vitest writes frontend dump to a file, Go
// test reads it + compares to backend nodekind.ExportJSON() output.
//
// CRITICAL: any field added here MUST be added to backend ExportedSpec too,
// AND the parity test's skip list updated if it's not meant to match across.

import { allSpecs } from './registry'
import type { PinType } from './index'

export interface ExportedSpec {
  kind: string
  group: string
  /** null when empty (matches Go's `omitempty` + json.Marshal of nil slice). */
  execIn: string[] | null
  execOut: string[] | null
  hasExecOutFn: boolean
  /** null when empty (matches Go side). */
  dataIn: Record<string, PinType> | null
  hasDataInDynamicFn: boolean
  dataOut: Record<string, PinType> | null
  isPureData: boolean
  isVisualOnly: boolean
}

/**
 * exportJSON returns all registered specs as a stable JSON string,
 * sorted by kind ascending. Designed to byte-match backend nodekind.ExportJSON.
 */
export function exportJSON(): string {
  const out: ExportedSpec[] = allSpecs().map((s) => ({
    kind: s.kind,
    group: s.group,
    execIn: s.execIn.length > 0 ? s.execIn : null,
    execOut: s.execOut.length > 0 ? s.execOut : null,
    hasExecOutFn: !!s.execOutFn,
    dataIn: Object.keys(s.dataIn).length > 0 ? sortKeys(s.dataIn) : null,
    hasDataInDynamicFn: !!s.dataInDynamicFn,
    dataOut: Object.keys(s.dataOut).length > 0 ? sortKeys(s.dataOut) : null,
    isPureData: !!s.isPureData,
    isVisualOnly: !!s.isVisualOnly,
  }))
  out.sort((a, b) => (a.kind < b.kind ? -1 : a.kind > b.kind ? 1 : 0))
  // JSON.stringify with 2-space indent matches Go's json.MarshalIndent("", "  ").
  // Go marshals maps with keys sorted alphabetically; we mirror via sortKeys above.
  return JSON.stringify(out, null, 2)
}

/**
 * sortKeys returns a new object with keys sorted alphabetically — necessary
 * because Go's encoding/json marshals maps with sorted keys, but
 * JavaScript's JSON.stringify uses insertion order. Without this, FE and BE
 * dumps diff on map key order even with identical content.
 */
function sortKeys<T>(obj: Record<string, T>): Record<string, T> {
  const result: Record<string, T> = {}
  for (const k of Object.keys(obj).sort()) result[k] = obj[k]
  return result
}
